From 64c162af271879ead0adb7f1fee4820a7cc7d7a1 Mon Sep 17 00:00:00 2001
From: Mao Minkai <maominkai@wxiat.com>
Date: Tue, 7 Jun 2022 17:42:05 +0800
Subject: [PATCH 3002/4585] [Backport] sw64: add clone3 syscall support

Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I56XYC

Reference: https://gitee.com/openeuler/kernel/commit/5cb6b3fe8eca448cf175f380271a852df1c59603

--------------------------------

The implementation of clone3 on sw64 is a bit different from the
generic one.

Signed-off-by: Mao Minkai <maominkai@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Acked-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>

Signed-off-by: Wang Hai <wanghai38@huawei.com>
---
 arch/sw_64/include/asm/unistd.h        | 1 +
 arch/sw_64/kernel/entry.S              | 1 +
 arch/sw_64/kernel/syscalls/syscall.tbl | 2 +-
 3 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/arch/sw_64/include/asm/unistd.h b/arch/sw_64/include/asm/unistd.h
index c1778adf4fba..b8854163d96f 100644
--- a/arch/sw_64/include/asm/unistd.h
+++ b/arch/sw_64/include/asm/unistd.h
@@ -22,5 +22,6 @@
 #define __ARCH_WANT_SYS_CLONE
 #define __ARCH_WANT_SYS_SOCKETCALL
 #define __ARCH_WANT_SYS_SIGPROCMASK
+#define __ARCH_WANT_SYS_CLONE3
 
 #endif /* _ASM_SW64_UNISTD_H */
diff --git a/arch/sw_64/kernel/entry.S b/arch/sw_64/kernel/entry.S
index 753eb31a76c6..c12487e52141 100644
--- a/arch/sw_64/kernel/entry.S
+++ b/arch/sw_64/kernel/entry.S
@@ -659,6 +659,7 @@ sw64_\name:
 fork_like fork
 fork_like vfork
 fork_like clone
+fork_like clone3
 
 	.align 4
 	.globl sys_sigreturn
diff --git a/arch/sw_64/kernel/syscalls/syscall.tbl b/arch/sw_64/kernel/syscalls/syscall.tbl
index 6026c637d923..472957792bd2 100644
--- a/arch/sw_64/kernel/syscalls/syscall.tbl
+++ b/arch/sw_64/kernel/syscalls/syscall.tbl
@@ -289,7 +289,7 @@
 279	common	fsmount				sys_fsmount
 280	common	fspick				sys_fspick
 281	common	pidfd_open			sys_pidfd_open
-282	common	clone3				sys_clone3
+282	common	clone3				sw64_clone3
 283	common	close_range			sys_close_range
 284	common	openat2				sys_openat2
 285	common	pidfd_getfd			sys_pidfd_getfd
-- 
2.17.1

